### PR TITLE
Add upgrade form v4 to v5.1.0

### PR DESCRIPTION
**Pull request by wivwiv**

### CODE REVIEW

File: zh_CN/deploy/upgrade-from-v4-to-v51.md [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+
+# 5.1
+{allow, {username, {re, "^dashboard$"}}, subscribe, ["$SYS/#"]}.
+{allow, {ipaddr, "127.0.0.1"}, all, ["$SYS/#", "#"]}.
+```
+
+#### 内置数据库 (Mnesia) 变动
+
+1. 为方便理解，该数据源由 Mnesia 更名为内置数据库(Built-in Database)；
+2. 数据格式与数据操作 REST API 有变动，新的 API 是 `/authorization/sources/built_in_database/rules{/clients,/users}`。
```
Comment (Meggielqk) [inline zh_CN/deploy/upgrade-from-v4-to-v51.md:307]: 这部分的英文和中文不完全一致，需要统一吗？

**Comment by Meggielqk** [comment]
"What's new" (new-features.md) 这个文件也需要更新一版吧？

File: zh_CN/deploy/upgrade-from-v4-to-v51.md [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+
+## 代理订阅（从数据库中获取订阅关系）
+
+从 5.0.0 版本开始，EMQX 不再提供基于外置数据库的[代理订阅](https://docs.emqx.com/zh/enterprise/v4.4/rule/get_subs_from_redis.html)（从数据库中获取订阅关系）功能。
+
+{% endemqxee %}
+
+## WebHook
+
+4.x 版本中的 WebHook 插件 (emqx_web_hook) 已被移除，请使用数据集成中的 HTTP Server 数据桥接功能替代。
```
Comment (Meggielqk) [inline zh_CN/deploy/upgrade-from-v4-to-v51.md:425]: 现在是叫 “HTTP Service” 吧？

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+}
+```
+
+## Rule Engine
+
+Rule SQL is fully compatible with 4.x syntax, but the actions under the rule are split into built-in actions (republish, console) and data bridges (HTTP Server, MQTT Bridge).
+
+## WebHook
+
+The WebHook plugin (emqx_web_hook) has been removed. Use the HTTP Server data bridge in Data Integration instead.
```
Comment (Meggielqk) [inline en_US/deploy/upgrade-from-v4-to-v51.md:410]: HTTP Service？

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+By default, the following listeners are no longer provided:
+
+| Listeners            | Description                        |
+| -------------------- | ---------------------------------- |
+| MQTT-TCP 11883       | Listeners for backend applications |
+| Management-HTTP 8081 | For REST API, merged to port 18083 |
+
+## Plugins
+
+The previous official plugins have been migrated to EMQX as built-in functions. Custom plugins developed for version 4.x need to be adapted before they can be used in EMQX 5.1. The following is a comparison table between official plugins and existing functions:
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The previous official plugins have been migrated to EMQX as built-in functions. Custom plugins developed for version 4.x need to be adapted before they can be used in EMQX 5.1. The following is a comparison table between official plugins and built-in features.
```

File: zh_CN/deploy/upgrade-from-v4-to-v51.md (hunk 1 above)
Comment (wivwiv) [reply zh_CN/deploy/upgrade-from-v4-to-v51.md:307]: 以中文为准，中文是最新的。

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 5]
```diff
@@ -0,0 +1,488 @@
+# Migrate from V4
+
+This section provides guidelines for migrating from EMQX 4.x to 5.1.0.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
This section provides guidelines for migrating from EMQX 4.4 to 5.1 or newer.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+EMQX 5.1.0 is not backward compatible with 4.x in management APIs and clustering APIs. However, most of the functionality is not significantly changed.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
EMQX 5 is not backward compatible with 4.x in management APIs and clustering APIs. However, most of the functionality is not significantly changed.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+<!-- To learn about the new features we added to EMQX 5.1, check [New Features](../getting-started/new-features.md). -->
+
+## Log
+
+Compared with 4.x, the most significant change in the log is the format. In 4.x, the log text is primarily plain, aiming for good human readability. However, starting from 5.1, we moved to structured logging without losing readability.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
Compared with 4.x, the most significant change in the log is the format. In 4.x, the log text is primarily plain, aiming for good human readability. Starting from 5, logging style is changed to be more structured, so it is more machine (log indexer).
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+For instance, most log fields use underscores as word separators, making them more search-friendly. This format also helps log indexing tools to index the logs more effectively.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The `msg` field value is a snake_case message, making them more search-friendly. This format also helps log indexing tools to index the logs more effectively, for instance:
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+| **Users/Alarms**                 |                                             |                      |                              |
+| `GET /users`                     | `GET /users`                                | Partially compatible |                              |
+| `GET /alarms{/activated}`        | `GET /alarms?activated={true,false}`        | Incompatible         |                              |
+| `GET /alarms{/deactivated}`      | `GET /alarms?activated={true,false}`        | Incompatible         |                              |
+
+## Auth/ACL
+
+**Function Entry**
+
+The Auth and ACL plugins (emqx_auth_*) have been removed, and the corresponding functions are built into EMQX. Users can configure them through Dashboard or configuration files.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+
+Auth is renamed to **Authentication**, and ACL is renamed **Authorization**. Now they are independent of each other.
+
+**Data Migration**
+
+We have kept the authentication methods and supported data sources from version 4.x, with only a few changes to the way they are used. For most Authenticator/Authorizer, you can continue to use the 4.x database when upgrading to version 5.1 without having to migrate existing data.
+
+### Order of authentication/authorization chain
+
+When multiple authenticators or authorization checkers are enabled simultaneously, the checks are no longer performed according to the startup order but a fixed configuration order. The execution order can be adjusted in the configuration file and Dashboard.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
When multiple authenticators or authorization checkers are enabled simultaneously, the checks are no longer performed according to the startup order but a fixed order per configuration. The order can be adjusted from Dashboard.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+  {
+    ...
+    mechanism = "password_based"
+    backend = "mysql"
+    query = "SELECT password_hash, salt FROM mqtt_user where username = ${username} LIMIT 1"
+  }
+]
+```
+
+### Auth
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
### MQTT Client Authentication
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+| `GET /nodes`                     | `GET /nodes`                                | Partially compatible |                              |
+| `GET /brokers`                   | -                                           | Incompatible         | merged to `GET /nodes`       |
+| `GET /stats`                     | `GET /stats`                                | Partially compatible |                              |
+| `GET /metrics`                   | `GET /metrics`                              | Partially compatible |                              |
+| **Users/Alarms**                 |                                             |                      |                              |
+| `GET /users`                     | `GET /users`                                | Partially compatible |                              |
+| `GET /alarms{/activated}`        | `GET /alarms?activated={true,false}`        | Incompatible         |                              |
+| `GET /alarms{/deactivated}`      | `GET /alarms?activated={true,false}`        | Incompatible         |                              |
+
+## Auth/ACL
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
## MQTT Client Authentication and Authorization
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+
+**Function Entry**
+
+The Auth and ACL plugins (emqx_auth_*) have been removed, and the corresponding functions are built into EMQX. Users can configure them through Dashboard or configuration files.
+
+**Change of concept**
+
+Auth is renamed to **Authentication**, and ACL is renamed **Authorization**. Now they are independent of each other.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
Auth is referred to as **Authentication**, and ACL is also referred to as **Authorization**.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+    query = "SELECT password_hash, salt FROM mqtt_user where username = ${username} LIMIT 1"
+  }
+]
+```
+
+### Auth
+
+#### Remove the Anonymous Mechanism
+
+Remove the `allow_anonymous` configuration item. All client connections are allowed by default. If **add and enable** any authenticator, EMQX will perform an authentication check for all new connections.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The `allow_anonymous` configuration is now deleted. All clients are allowed to connect by default. If **add and enable** any authenticator, EMQX will try to authenticate the clients.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+
+When a client matches the authentication data in all authenticators, the connection is rejected.
+
+Remove the `bypass_auth_plugins` configuration item. When a listener needs to skip authentication, it can be set through the `listeners.{type}.{name}.enable_authn = true | false` configuration item.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The `bypass_auth_plugins` configuration is also deleted. When one wants to allow all clients to connect without authentication, they may set `listeners.{type}.{name}.enable_authn = false`.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 16]
```diff
@@ -0,0 +1,488 @@
...
+]
+```
+
+### Auth
+
+#### Remove the Anonymous Mechanism
+
+Remove the `allow_anonymous` configuration item. All client connections are allowed by default. If **add and enable** any authenticator, EMQX will perform an authentication check for all new connections.
+
+When a client matches the authentication data in all authenticators, the connection is rejected.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
After traversing the configured authentication chain, if none of the authenticator in the chain can decide that this client is allowed to connect, then the connection is rejected.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 17]
```diff
@@ -0,0 +1,488 @@
...
+
+Remove the `bypass_auth_plugins` configuration item. When a listener needs to skip authentication, it can be set through the `listeners.{type}.{name}.enable_authn = true | false` configuration item.
+
+#### Built-in database (Mnesia)
+
+1. Mnesia is renamed to the built-in database;
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
1. Mnesia is now referred to as the "built-in" database;
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 18]
```diff
@@ -0,0 +1,488 @@
...
+2. Only two search methods are available: username-based or clientid-based;
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
2. Two search methods are available: username-based or clientid-based, but mixing is no longer supported.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 19]
```diff
@@ -0,0 +1,488 @@
...
+3. The data format and REST API have changed. For more information, refer to `POST /authentication/{id}/users`.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
3. The REST APIs to manage the authentication data records are changed. For more information, refer to the API doc for `POST /authentication/{id}/users`.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 20]
```diff
@@ -0,0 +1,488 @@
...
+
+#### HTTP
+
+1. Use the JSON fields inside the response body instead of the HTTP response status codes to identify the authentication result;
+2. Remove standalone super-user request. The super-user identity of the client is established through authentication in the response body.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md:202]:
```suggestion
2. Removed standalone super-user request. The super-user identity of the client is established through authentication in the response body.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 21]
```diff
@@ -0,0 +1,488 @@
...
+{
+  "result": "deny"
+}
+```
+
+## Rule Engine
+
+Rule SQL is fully compatible with 4.x syntax, but the actions under the rule are split into built-in actions (republish, console) and data bridges (HTTP Server, MQTT Bridge).
+
+## WebHook
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
## HTTP Server
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 22]
```diff
@@ -0,0 +1,488 @@
...
+
+The WebHook plugin (emqx_web_hook) has been removed. Use the HTTP Server data bridge in Data Integration instead.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The WebHook plugin (`emqx_web_hook`) is converted to a native feature, and it is now referred to as "HTTP Server" bridge.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 23]
```diff
@@ -0,0 +1,488 @@
...
+
+## MQTT Bridge
+
+The MQTT bridge plugin (emqx_bridge_mqtt) has been removed. Use the MQTT Sink and MQTT Source data bridge in Data Integration instead.
+
+{% emqxee %}
+
+## Offline messages
+
+The [offline messages](https://docs.emqx.com/en/enterprise/v4.4/rule/offline_msg_to_redis.html) provided in EMQX 4.x are based on an external database. EMQX plans to provide native offline messages (based on the built-in database) in future versions, so the offline messages for the external database have been removed from version 5.0.0.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The [offline messages](https://docs.emqx.com/en/enterprise/v4.4/rule/offline_msg_to_redis.html) provided in EMQX 4.x are based on an external database. EMQX plans to provide native offline messages (based on the built-in database) in future versions, so the offline messages for the external database is no longer supported in version 5.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 24]
```diff
@@ -0,0 +1,488 @@
...
+
+## Auto subscription (Server side subscriptions)
+
+As of version 5.0.0, EMQX no longer provides [Auto subscription](https://docs.emqx.com/en/enterprise/v4.4/rule/get_subs_from_redis.html) (server side subscriptions) based on an external database.
+
+{% endemqxee %}
+
+## Prometheus
+
+Old Prometheus plug-in (emqx_prometheus) has been removed. In version 5.1, Prometheus scraping endpoint is enabled by default, and no authentication is required to scrap the metrics.
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]:
```suggestion
The old plugin named `emqx_prometheus` has been converted to a native feature in version 5. Prometheus scraping endpoint is enabled by default, and no authentication is required to scrap the metrics.
```

File: en_US/deploy/upgrade-from-v4-to-v51.md [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+Clients of other protocols than MQTT (LwM2M, CoAP, STOMP, MQTT-SN) are no longer listed as MQTT clients on the dashboard **Connections** page and in the `GET /clients` API.
+They can be found in **Extentions** -> **Gateways** section or listed with `GET /gateway/{name}/clients` API.
+
+{% emqxce %}
+
+## Telemetry
+
+The telemetry plugin (emqx_telemetry) has been removed. Please configure it through the `telemetry {}` configuration item or in the Dashboard **System Settings** -> **Settings** page.
+
+{% endemqxce %}
```
Comment (zmstone) [inline en_US/deploy/upgrade-from-v4-to-v51.md]: in v5.0.26, emqx_telemetry is added back as an application, and there is not really much changed.
```suggestion
```

**Review by Meggielqk — APPROVED** [review]

**Pull request merged**